apps/bfd-pipeline/bfd-pipeline-rda-grpc-server/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/ConfigLoader.java

```diff
@@ -0,0 +1,253 @@
+package gov.cms.bfd.pipeline.rda.grpc.sink;
+
+import com.google.common.base.Strings;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.Reader;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.function.Function;
+
+/** Simple class to abstract loading configuration values from multiple sources. */
+public class ConfigLoader {
+  private final Function<String, String> source;
```

Comment:
I was a bit thrown off by the name 'source' here and below. This does not really represent the source itself but rather the mapping function that the source gives rise to, and for this particular member it is a composite mapping that arises from potentially multiple sources, right? A javadoc on this member regardless of what we decide to call it would be appreciated.

Also, what were the factors that lead to using a Function<> instead of a Map<>? With the "new" mappings overwriting the "old" ones as sources are added I don't see anything that really requires the Function<> and it can make it more annoying to debug later if you want to print out the mapping in a debugger.

Reply:
The source of data for this loader is a function it can call to obtain a value using a key.  How that value is obtained doesn't matter to the ConfigLoader object.   So from the perspective of this class `source` actually is a source of data.  Certainly in some cases this source could be some sort of mapping that does extra stuff but generally it's more likely to be a closure of a collection and its getter method.

You are correct that the `source` is often a composite mapping but it doesn't have to be.  It could just be `map::get` or `System::getenv`.

I agree a comment is needed.  I'll add that.

Using a Function means the source can be literally anything and values can be obtained lazily.  Using a Map would require creating a new data structure up front and passing it to the ConfigLoader.   For example we don't want to create a Map of all possible environment variables and then add all system properties to that map.  Most of that work would be wasted since the ConfigLoader would only care about a few of the keys in the resulting Map.  What we need is just a way to grab only the values it needs and only when it needs them.  A lambda is great for that.